mleap-core/src/main/scala/ml/combust/mleap/core/types/Casting.scala

```diff
@@ -245,32 +253,88 @@ object Casting {
+                      case _ => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                    }
+                  }
                 case BasicType.Short =>
                   val cc = c.asInstanceOf[(Any) => Short]
-                  (v: Any) => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                  (v: Any) => {
+                    v match {
+                      case _: SparkVector => sparkVectorToMLeapTensor(v.asInstanceOf[SparkVector]).mapValues(cc)
+                      case _ => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                    }
+                  }
                 case BasicType.Int =>
                   val cc = c.asInstanceOf[(Any) => Int]
-                  (v: Any) => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                  (v: Any) => {
+                    v match {
+                      case _: SparkVector => sparkVectorToMLeapTensor(v.asInstanceOf[SparkVector]).mapValues(cc)
+                      case _ => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                    }
+                  }
                 case BasicType.Long =>
                   val cc = c.asInstanceOf[(Any) => Long]
-                  (v: Any) => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                  (v: Any) => {
+                    v match {
+                      case _: SparkVector => sparkVectorToMLeapTensor(v.asInstanceOf[SparkVector]).mapValues(cc)
+                      case _ => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                    }
+                  }
                 case BasicType.Float =>
                   val cc = c.asInstanceOf[(Any) => Float]
-                  (v: Any) => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                  (v: Any) => {
+                    v match {
+                      case _: SparkVector => sparkVectorToMLeapTensor(v.asInstanceOf[SparkVector]).mapValues(cc)
+                      case _ => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                    }
+                  }
                 case BasicType.Double =>
                   val cc = c.asInstanceOf[(Any) => Double]
-                  (v: Any) => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                  (v: Any) => {
+                    v match {
+                      case _: SparkVector => sparkVectorToMLeapTensor(v.asInstanceOf[SparkVector]).mapValues(cc)
+                      case _ => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                    }
+                  }
                 case BasicType.String =>
                   val cc = c.asInstanceOf[(Any) => String]
-                  (v: Any) => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                  (v: Any) => {
+                    v match {
+                      case _: SparkVector => sparkVectorToMLeapTensor(v.asInstanceOf[SparkVector]).mapValues(cc)
+                      case _ => v.asInstanceOf[Tensor[_]].mapValues(cc)
+                    }
```

Comment:
this code is copy-pasted quite a bit, could make it a helper?

Reply:
I couldn't get the generic function definition to behave, so I'm just going to leave it as-is for now